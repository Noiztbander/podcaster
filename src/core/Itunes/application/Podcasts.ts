import type { IPodcast } from "src/core/Itunes/domain/Podcasts";
import { IPodcastsRepository } from "src/core/Itunes/infra/PodcastsRepository";

export interface IPodcastsGetter {
  get(): Promise<IPodcast[]>;
}

export class PodcastsGetter implements IPodcastsGetter {
  constructor(private readonly repository: IPodcastsRepository) {}

  private fetchPodcasts = async (): Promise<IPodcast[]> => {
    const resp = await this.repository.get();

    return resp.feed.entry;
  };

  async get(): Promise<IPodcast[]> {
    return this.fetchPodcasts();
  }
}
