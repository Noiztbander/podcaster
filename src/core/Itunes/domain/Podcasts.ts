export enum WrapperTypeEnum {
  TRACK = 'track',
  EPISODE = 'podcastEpisode',
}

interface ITunesImage {
  label: string;
  attributes: { height: string };
}

export interface IpodcastDetail {
  wrapperType: WrapperTypeEnum.TRACK;
  collectionId: string;
}

export interface IEpisode {
  wrapperType: WrapperTypeEnum.EPISODE;
  trackTimeMillis: number;
  releaseDate: string;
  trackName: string;
  episodeUrl: string;
  trackId: number;
  description: string;
}

export interface IPodcastDetailRestApi {
  results: (IpodcastDetail | IEpisode)[];
}

export interface IPodcastRestApi {
  'im:name': { label: string };
  'im:image': ITunesImage[];
  'im:artist': { label: string };
  id: { attributes: { 'im:id': string } };
  summary: { label: string };
}

export interface Ipodcast {
  name: string;
  image: string;
  artist: string;
  id: string;
  summary: string;
}
