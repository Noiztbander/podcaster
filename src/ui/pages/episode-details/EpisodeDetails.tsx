import styles from './EpisodeDetails.module.scss';
import Header from '@/ui/components/header/Header';

import PodcastDetailItem from '@/ui/components/podcast-detail-item/PodcastDetailItem';
import useLocalStorage from '@/ui/hook/useLocalStorage';

export default function EpisodeDetails() {
  const { episode } = useLocalStorage();

  return (
    <main>
      <Header />
      <div className={styles.PageContainer}>
        <PodcastDetailItem />
        <section className={styles.episodeSection}>
          <div className={styles.title}>{episode.trackName}</div>

          <div>
            <textarea readOnly className={styles.textArea} defaultValue={episode.description} />
          </div>
          <div className={styles.audio}>
            <audio controls autoPlay preload="auto">
              <source src={episode.episodeUrl} type="audio/mpeg" />
            </audio>
          </div>
        </section>
      </div>
    </main>
  );
}
