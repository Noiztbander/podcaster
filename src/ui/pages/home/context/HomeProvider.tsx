import { useMemo } from "react";
import type { HomeProviderProps } from "../Home.types";
import HomeContext from "./HomeContext";

const HomeProvider = ({ podcasts, children }: HomeProviderProps) => {
  const watchProps = useMemo(() => ({ podcasts }), [podcasts]);

  return (
    <HomeContext.Provider value={watchProps}>{children}</HomeContext.Provider>
  );
};

export default HomeProvider;
