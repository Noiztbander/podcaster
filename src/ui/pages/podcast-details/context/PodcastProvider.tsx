import { useMemo } from 'react';
import type { IPodcastProviderProps } from '../PodcastDetails.types';
import PodcastContext from './PodcastContext';

const PodcastProvider = ({ episodesList, children }: IPodcastProviderProps) => {
  const watchProps = useMemo(() => ({ episodesList }), [episodesList]);

  return <PodcastContext.Provider value={watchProps}>{children}</PodcastContext.Provider>;
};

export default PodcastProvider;
