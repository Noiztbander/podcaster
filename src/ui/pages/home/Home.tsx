import { Ipodcast } from '@/core/Itunes/domain/Podcasts';
import Header from '@/ui/components/header/Header';
import useInputFilter from '@/ui/hook/useInputFilter';
import { localStorageEnum } from '@/ui/types/localStorage.types';
import { updateStorage } from '@/ui/utils/storage.utils';
import { useContext, useEffect, useState } from 'react';
import InputFilter from './components/podcast-card/input-filter/InputFilter';
import PodcastCard from './components/podcast-card/PodcastCard';

import HomeContext from './context/HomeContext';
import styles from './home.module.scss';
import { IHomeContext } from './Home.types';

export default function Home() {
  const { podcasts = [] } = useContext(HomeContext) as IHomeContext;
  const [query, setQuery] = useState('');

  useEffect(() => {
    updateStorage({ [localStorageEnum.TOPPODCASTS]: podcasts });
  }, [podcasts]);

  const { filteredPodcasts } = useInputFilter({ query });

  return (
    <main>
      <Header />
      <InputFilter setQuery={setQuery} podcastQuantity={filteredPodcasts.length} />
      <section className={styles.podcastSection}>
        {filteredPodcasts.map((podcast: Ipodcast, index) => (
          <PodcastCard key={index} podcast={podcast} />
        ))}
      </section>
    </main>
  );
}
