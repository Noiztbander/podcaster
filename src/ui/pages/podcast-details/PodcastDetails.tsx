import { useContext, useEffect } from 'react';
import PodcastContext from './context/PodcastContext';
import { IPodcastContext } from './PodcastDetails.types';
import styles from './PodcastDetails.module.scss';
import Header from '@/ui/components/header/Header';
import { updateStorage } from '@/ui/utils/storage.utils';
import { localStorageEnum } from '@/ui/types/localStorage.types';
import useLocalStorage from '@/ui/hook/useLocalStorage';
import EpisodeRowItem from './components/episode-row-item/EpisodeRowitem';
import { IEpisode } from '@/core/Itunes/domain/Podcasts';
import PodcastDetailItem from '../../components/podcast-detail-item/PodcastDetailItem';

export default function PodcastDetails() {
  const { episodesList } = useContext(PodcastContext) as IPodcastContext;

  useEffect(() => {
    updateStorage({ [localStorageEnum.EPISODES]: episodesList });
  }, [episodesList]);

  const { episodes } = useLocalStorage();

  return (
    <main>
      <Header />
      <div className={styles.PageContainer}>
        <PodcastDetailItem />
        <section className={styles.episodesListSection}>
          <div className={styles.titleList}>
            <h3>Episodes: {`${episodes.length}`}</h3>
          </div>
          <div className={styles.episodesList}>
            <div className={styles.titles}>
              <h5>Title</h5>
              <h5>Date</h5>
              <h5>Duration</h5>
            </div>
            <ul>
              {episodes.map((episode: IEpisode, index) => (
                <EpisodeRowItem episode={episode} key={index} />
              ))}
            </ul>
          </div>
        </section>
      </div>
    </main>
  );
}
