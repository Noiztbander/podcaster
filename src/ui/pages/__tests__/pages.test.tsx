import { render } from '@testing-library/react';
import EpisodeDetails from '../episode-details/EpisodeDetails';
import HomeProvider from '../home/context/HomeProvider';
import Home from '../home/Home';
import PodcastProvider from '../podcast-details/context/PodcastProvider';
import PodcastDetails from '../podcast-details/PodcastDetails';
import podcastMock from './podcasts.json';
import episodesMock from './episodes.json';
import { IEpisode } from '@/core/Itunes/domain/Podcasts';

jest.mock('next/router', () => ({
  useRouter: jest.fn(),
}));

describe('<Home />', () => {
  it('should render the component correctly with default values', () => {
    const { asFragment } = render(
      <HomeProvider podcasts={[]}>
        <Home />
      </HomeProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the component correctly with some value', () => {
    const { asFragment } = render(
      <HomeProvider podcasts={podcastMock}>
        <Home />
      </HomeProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});

describe('<PodcastDetails />', () => {
  it('should render the component correctly with default values', () => {
    const { asFragment } = render(
      <PodcastProvider episodesList={[]}>
        <PodcastDetails />
      </PodcastProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the component correctly with some value', () => {
    const { asFragment } = render(
      <PodcastProvider episodesList={episodesMock as IEpisode[]}>
        <PodcastDetails />
      </PodcastProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});

describe('<<EpisodeDetails /> />', () => {
  it('should render the component correctly with default values', () => {
    const { asFragment } = render(<EpisodeDetails />);

    expect(asFragment()).toMatchSnapshot();
  });
});
