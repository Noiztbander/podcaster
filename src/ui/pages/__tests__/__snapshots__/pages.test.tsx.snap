// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<<EpisodeDetails /> /> should render the component correctly with default values 1`] = `
<DocumentFragment>
  <main>
    <section
      class="podcasterSection"
    >
      <a
        href="/"
      >
        <h3
          class="titlePodcaster"
        >
          Podcaster
        </h3>
      </a>
      <div
        class="spinnerContainer"
      >
        <div
          class="loadingSpinner"
          id="header-loading-spinner"
        >
          <div />
          <div />
          <div />
          <div />
        </div>
      </div>
    </section>
    <div
      class="PageContainer"
    >
      <section
        class="podcastSection"
      >
        <a
          class="imagePodcast"
          href="/podcast/undefined"
        >
          <img />
        </a>
        <a
          class="titlePodcast"
          href="/podcast/undefined"
        >
          <strong>
            <h4 />
          </strong>
          <h5>
            by 
          </h5>
        </a>
        <div
          class="summaryPodcast"
        >
          <strong>
            <h4>
              Description:
            </h4>
          </strong>
          <p />
        </div>
      </section>
      <section
        class="episodeSection"
      >
        <div
          class="title"
        />
        <div>
          <textarea
            class="textArea"
            readonly=""
          />
        </div>
        <div
          class="audio"
        />
      </section>
    </div>
  </main>
</DocumentFragment>
`;

exports[`<Home /> should render the component correctly with default values 1`] = `
<DocumentFragment>
  <main>
    <section
      class="podcasterSection"
    >
      <a
        href="/"
      >
        <h3
          class="titlePodcaster"
        >
          Podcaster
        </h3>
      </a>
      <div
        class="spinnerContainer"
      >
        <div
          class="loadingSpinner"
          id="header-loading-spinner"
        >
          <div />
          <div />
          <div />
          <div />
        </div>
      </div>
    </section>
    <section
      class="filterSection"
    >
      <div
        class="podcastCounter"
      >
        <h3>
          0
        </h3>
      </div>
      <label
        class="label"
        for="filter-podcasts"
      >
        <input
          id="filter-podcasts"
          name="filter-podcasts"
          placeholder="Filter podcasts..."
          type="text"
        />
      </label>
    </section>
    <section
      class="podcastSection"
    />
  </main>
</DocumentFragment>
`;

exports[`<Home /> should render the component correctly with some value 1`] = `
<DocumentFragment>
  <main>
    <section
      class="podcasterSection"
    >
      <a
        href="/"
      >
        <h3
          class="titlePodcaster"
        >
          Podcaster
        </h3>
      </a>
      <div
        class="spinnerContainer"
      >
        <div
          class="loadingSpinner"
          id="header-loading-spinner"
        >
          <div />
          <div />
          <div />
          <div />
        </div>
      </div>
    </section>
    <section
      class="filterSection"
    >
      <div
        class="podcastCounter"
      >
        <h3>
          2
        </h3>
      </div>
      <label
        class="label"
        for="filter-podcasts"
      >
        <input
          id="filter-podcasts"
          name="filter-podcasts"
          placeholder="Filter podcasts..."
          type="text"
        />
      </label>
    </section>
    <section
      class="podcastSection"
    >
      <div
        class="podcastCard"
      >
        <a
          class="bg"
          href="/podcast/1535809341"
        >
          <div
            class="image"
          >
            <img
              alt="The Joe Budden Podcast"
              src="https://is1-ssl.mzstatic.com/image/thumb/Podcasts113/v4/f2/21/fa/f221fabd-017f-5125-633b-f1fe4f39802a/mza_182995249085044287.jpg/170x170bb.png"
            />
          </div>
          <div
            class="titles"
          >
            <h4>
              THE JOE BUDDEN PODCAST
            </h4>
            <h5>
              The Joe Budden Network
            </h5>
          </div>
        </a>
      </div>
      <div
        class="podcastCard"
      >
        <a
          class="bg"
          href="/podcast/1481398762"
        >
          <div
            class="image"
          >
            <img
              alt="Dolly Parton's America"
              src="https://is2-ssl.mzstatic.com/image/thumb/Podcasts115/v4/58/b8/16/58b8165d-d674-16cd-3822-36d8acad0151/mza_5834995761907664558.jpg/170x170bb.png"
            />
          </div>
          <div
            class="titles"
          >
            <h4>
              DOLLY PARTON'S AMERICA
            </h4>
            <h5>
              WNYC Studios & OSM Audio
            </h5>
          </div>
        </a>
      </div>
    </section>
  </main>
</DocumentFragment>
`;

exports[`<PodcastDetails /> should render the component correctly with default values 1`] = `
<DocumentFragment>
  <main>
    <section
      class="podcasterSection"
    >
      <a
        href="/"
      >
        <h3
          class="titlePodcaster"
        >
          Podcaster
        </h3>
      </a>
      <div
        class="spinnerContainer"
      >
        <div
          class="loadingSpinner"
          id="header-loading-spinner"
        >
          <div />
          <div />
          <div />
          <div />
        </div>
      </div>
    </section>
    <div
      class="PageContainer"
    >
      <section
        class="podcastSection"
      >
        <a
          class="imagePodcast"
          href="/podcast/undefined"
        >
          <img />
        </a>
        <a
          class="titlePodcast"
          href="/podcast/undefined"
        >
          <strong>
            <h4 />
          </strong>
          <h5>
            by 
          </h5>
        </a>
        <div
          class="summaryPodcast"
        >
          <strong>
            <h4>
              Description:
            </h4>
          </strong>
          <p />
        </div>
      </section>
      <section
        class="episodesListSection"
      >
        <div
          class="titleList"
        >
          <h3>
            Episodes: 0
          </h3>
        </div>
        <div
          class="episodesList"
        >
          <div
            class="titles"
          >
            <h5>
              Title
            </h5>
            <h5>
              Date
            </h5>
            <h5>
              Duration
            </h5>
          </div>
          <ul />
        </div>
      </section>
    </div>
  </main>
</DocumentFragment>
`;

exports[`<PodcastDetails /> should render the component correctly with some value 1`] = `
<DocumentFragment>
  <main>
    <section
      class="podcasterSection"
    >
      <a
        href="/"
      >
        <h3
          class="titlePodcaster"
        >
          Podcaster
        </h3>
      </a>
      <div
        class="spinnerContainer"
      >
        <div
          class="loadingSpinner"
          id="header-loading-spinner"
        >
          <div />
          <div />
          <div />
          <div />
        </div>
      </div>
    </section>
    <div
      class="PageContainer"
    >
      <section
        class="podcastSection"
      >
        <a
          class="imagePodcast"
          href="/podcast/undefined"
        >
          <img />
        </a>
        <a
          class="titlePodcast"
          href="/podcast/undefined"
        >
          <strong>
            <h4 />
          </strong>
          <h5>
            by 
          </h5>
        </a>
        <div
          class="summaryPodcast"
        >
          <strong>
            <h4>
              Description:
            </h4>
          </strong>
          <p />
        </div>
      </section>
      <section
        class="episodesListSection"
      >
        <div
          class="titleList"
        >
          <h3>
            Episodes: 1
          </h3>
        </div>
        <div
          class="episodesList"
        >
          <div
            class="titles"
          >
            <h5>
              Title
            </h5>
            <h5>
              Date
            </h5>
            <h5>
              Duration
            </h5>
          </div>
          <ul>
            <li>
              <a
                class="liItem"
                href="/podcast//episode/1000598416768"
              >
                <p
                  class="title"
                >
                  Incredible Work
                </p>
                <p>
                  07/02/2023
                </p>
                <p>
                  42:04
                </p>
              </a>
            </li>
          </ul>
        </div>
      </section>
    </div>
  </main>
</DocumentFragment>
`;
